### ISSUE TITLE
Unable to obtain access token

### ISSUE THREAD

**Issue by afriedsam** [mmohades/Venmo#86]
Running the following code:
`from venmo_api import Client
>>>
>>> at = Client.get_access_token(username='*****************', password='***********')`

getting the following response:
`venmo_api.models.exception.HttpCodeError: HTTP Status code is invalid. Could not make the request because -> 400 Bad Request.
Error: {'error': {'title': "Couldn't Log In", 'message': 'Your email or password was incorrect.', 'links': None, 'code': 264}}`

My friend is having the same issue.

**Comment by robhannon** [comment]
Same here

**Comment by ahatzz11** [comment]
I tried to get a new auth token and ran into the same thing today with `Client.get_access_token`. I was able to obtain a valid bearer token from observing network calls from a web session and that worked great, but being able to obtain a new token from the python lib would be great.

**Comment by eddiew** [comment]
same here, as of today

**Comment by ahatzz11** [comment]
Since I ran into this I decided to do some digging to see if I could hack my way through this a bit. I was able to grab a valid api token from the network tab during a payment, but this python API still didn't seem to work due to a `403`. It seems like tokens created before whatever change venmo made are still considered valid, but I unfortunately revoked all the tokens on my account recently and it seems like they don't let you generate one of these tokens any more.

With this 'new' token it seems like there is now a change in their payment flow - I see a new call to `/api/eligibility` before the `api/payments` call. This API call returns:

```json
{
    "eligible": true,
    "eligibilityToken": "eyJ...",
    "fees": [
        {
            "productUri": "venmo:product:buyer_protection:f4de...",
            "appliedTo": "227...",
            "feePercentage": 0,
            "baseFeeAmount": 0,
            "calculatedFeeAmountInCents": 0,
            "feeToken": "eyJ..."
        }
    ]
}
```

The `api/payments` request then uses the `eligibilityToken`: 

```
{
    "targetUserDetails":
    {
        "userId": "227..."
    },
    "amountInCents": 1,
    "audience": "private",
    "note": "Test Transaction",
    "type": "pay",
    "fundingSourceID": "368...",
    "eligibilityToken": "eyJ..."
}
```

Based on the current [__send_or_request_money()](https://github.com/mmohades/Venmo/blob/0f0739dc8090d96073539722c57972df3ef315db/venmo_api/apis/payment_api.py#L195) function it doesn't seem like this token is being passed and is now something venmo requires. I've whipped up a quick local script to send in this new token (along with a `cookie` and `Csrf-Token` header) and was able to get this to work. I don't have the time to redo this python API to get it working, but figured I'd leave what I've found so far here in case some one else wants to tackle this issue.

**Comment by afriedsam** [comment]
If you log in while tracking the network requests you can also pull out both the deviceID and auth code from one of the requests and log in like this:

```
from venmo_api import Client
client = Client.get_access_token(username='*****************', password='***********', device_id="********")
```

or

```
from venmo_api import Client
client = Client(access_token="***************************")
```

To find device_id or access_token:
1. Log in while tracking the network requests and find the get request to https://account.venmo.com/ (it should be the top one)
2. Search for `graphqlHeaderConfig` in the response. Here you will find both the device_id and access_token which should work

I am trying to modify the code to pull this info from the request but running into some 403 errors--going to keep trying when I have some time. I don't think you need the eligibility token @ahatzz11.

**Comment by mmohades** [comment]
Thanks to everyone for flagging this with detailed info! I'm currently tied up, but I hope to look into this next weekend!

**Comment by mmohades** [comment]
It seems that Venmo may be actively flagging login attempts. If you use a trusted device-id with `get_access_token(username=..., password=..., device_id=TRUSTED_DEVICE_ID)`, it should work. I tested this and it worked for me.

What's interesting is that I removed my iPhone's device ID from the list of trusted devices on Venmo and logged out of the app. Upon attempting to log in via the API, I got the same 403 on the API. Then I tried to log in on my iPhone again, and I got the same error on my iPhone (see attached screenshot). 

<img src='https://github.com/mmohades/Venmo/assets/29171986/20f01f60-d94c-43ae-97a5-203201fa6d47' width='250'>

**Comment by jpjpjp** [comment]
> To find device_id or access_token:
> 
> 1. Log in while tracking the network requests and find the get request to https://account.venmo.com/ (it should be the top one)
> 2. Search for `graphqlHeaderConfig` in the response. Here you will find both the device_id and access_token which should work
> 

FWIW, I found this differently as I didn't see this particular request or any responses that included anything resembling "device_id". 

On the network request to `login`, I found in the response headers that it set a cookie called `v_id`.  I used the value of `v_id` as the `device_id` described above in the get_access_token() call.

**Comment by habersaat** [comment]
> Since I ran into this I decided to do some digging to see if I could hack my way through this a bit. I was able to grab a valid api token from the network tab during a payment, but this python API still didn't seem to work due to a `403`. It seems like tokens created before whatever change venmo made are still considered valid, but I unfortunately revoked all the tokens on my account recently and it seems like they don't let you generate one of these tokens any more.
> 
> With this 'new' token it seems like there is now a change in their payment flow - I see a new call to `/api/eligibility` before the `api/payments` call. This API call returns:
> 
> ```json
> {
>     "eligible": true,
>     "eligibilityToken": "eyJ...",
>     "fees": [
>         {
>             "productUri": "venmo:product:buyer_protection:f4de...",
>             "appliedTo": "227...",
>             "feePercentage": 0,
>             "baseFeeAmount": 0,
>             "calculatedFeeAmountInCents": 0,
>             "feeToken": "eyJ..."
>         }
>     ]
> }
> ```
> 
> The `api/payments` request then uses the `eligibilityToken`:
> 
> ```
> {
>     "targetUserDetails":
>     {
>         "userId": "227..."
>     },
>     "amountInCents": 1,
>     "audience": "private",
>     "note": "Test Transaction",
>     "type": "pay",
>     "fundingSourceID": "368...",
>     "eligibilityToken": "eyJ..."
> }
> ```
> 
> Based on the current [__send_or_request_money()](https://github.com/mmohades/Venmo/blob/0f0739dc8090d96073539722c57972df3ef315db/venmo_api/apis/payment_api.py#L195) function it doesn't seem like this token is being passed and is now something venmo requires. I've whipped up a quick local script to send in this new token (along with a `cookie` and `Csrf-Token` header) and was able to get this to work. I don't have the time to redo this python API to get it working, but figured I'd leave what I've found so far here in case some one else wants to tackle this issue.

@ahatzz11 I'm trying to recreate the same workaround but am having trouble retrieving the eligibility token (getting a 403 error each time). I'm only passing in the access_token in my post request header which I suspect is wrong... can I ask how you set the cookie and Csrf-Token header you mentioned?

**Comment by raycekar** [comment]
Was planning to try and maybe get a pull request in for a fix but found that when trying to log into venmo on any browser I am presented with a "Something went wrong. Try again later or contact customer support".  Script must have tried too many times and locked me out.  Anyone else have this for them and did you have to contact customer support?

**Comment by raycekar** [comment]
Sadly spend some hours on this and when trying to request through either of the following, I get a 404 error, invalid json and can't seem to figure out why:
https://account.venmo.com/api/eligibility
https://account.venmo.com/_next/data/PS8KMzrdGzuzFp7z0vH5i/en/pay.json

**Comment by raycekar** [comment]
Further digging, I see the HTTP version is HTTP/2 from the developer tools network tab.  From what I can see, the requests package only support HTTP/1.1 while another package, HTTPX, supports the new standard.  Could this be a possible reason for the 403/404 errors?

**Comment by mattmijo** [comment]
> It seems that Venmo may be actively flagging login attempts. If you use a trusted device-id with `get_access_token(username=..., password=..., device_id=TRUSTED_DEVICE_ID)`, it should work. I tested this and it worked for me.
> 
> What's interesting is that I removed my iPhone's device ID from the list of trusted devices on Venmo and logged out of the app. Upon attempting to log in via the API, I got the same 403 on the API. Then I tried to log in on my iPhone again, and I got the same error on my iPhone (see attached screenshot).
> 
> <img alt="" width="250" src="https://private-user-images.githubusercontent.com/29171986/320269444-20f01f60-d94c-43ae-97a5-203201fa6d47.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY2ODI1OTAsIm5iZiI6MTcxNjY4MjI5MCwicGF0aCI6Ii8yOTE3MTk4Ni8zMjAyNjk0NDQtMjBmMDFmNjAtZDk0Yy00M2FlLTk3YTUtMjAzMjAxZmE2ZDQ3LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI2VDAwMTEzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjMmQ5NmNhMGY0NDg4NGMyMDdhNGQ5YTgxYzkzODZhZTMwZDk3ZDIzZDdjZDM0MDA4Y2RmZThkNjViNDdkOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EtTF2BOcGoTzH34974M7rNX3TFbbztDa1T2XY4LHtlk">

I followed these steps but still results in a 403 error 

import requests
from venmo_api import Client
access_token = Client.get_access_token(username='blah', password='blah', device_id='longer blah')
print("My token:", access_token)
print(response.status_code)
print(response.json)

not sure what is happening here.

**Comment by KK2NJ** [comment]
Is there any update on this thread?

**Comment by KK2NJ** [comment]
> > Is there any update on this thread?
> 
> I tried, i failed... to get it to work.

There's a way to do it I can't remember how I did it but you gotta login into your venmo account + f12 + go to networking settings + enable no throttling + find your device id through the information there and then use the access_token code in the documentation

**Comment by KK2NJ** [comment]
> > > Is there any update on this thread?
> > 
> > 
> > I tried, i failed... to get it to work.
> 
> There's a way to do it I can't remember how I did it but you gotta login into your venmo account + f12 + go to networking settings + enable no throttling + find your device id through the information there and then use the access_token code in the documentation

Yeah I found it like the previous commenter said track your networking requests then search for "auth" and there should be "device_id" and "access_token" there you can take the access_token or just use Client.get_access_token with your login details and your device id in a IDE your choice.

**Comment by raycekar** [comment]
> > > Is there any update on this thread?
> > 
> > 
> > I tried, i failed... to get it to work.
> 
> There's a way to do it I can't remember how I did it but you gotta login into your venmo account + f12 + go to networking settings + enable no throttling + find your device id through the information there and then use the access_token code in the documentation

Yeah, I was able to get device ID and all, but it was the eligibility token portion that I was not able to implement to get it to work. Would you care to share the code you had updated.? Or I guess did you not make any updates to the Venmo code?

**Comment by rbrisita** [comment]
It was mentioned before but this is the steps I took to get a valid access token through this API:

1. Visit `https://venmo.com/account/sign-in`
2. Open DevTools
3. Manually sign into Venmo
4. Search for the `https://account.venmo.com/api/auth` request
5. Note the `deviceId` in the `Response`
6. Use that `deviceId` in your code when calling `Client.get_access_token`
7. The response will parrot the `device-id` and give a new `access_token`

**Comment by raycekar** [comment]
> It was mentioned before but this is the steps I took to get a valid access token through this API:
> 
>     1. Visit `https://venmo.com/account/sign-in`
> 
>     2. Open DevTools
> 
>     3. Manually sign into Venmo
> 
>     4. Search for the `https://account.venmo.com/api/auth` request
> 
>     5. Note the `deviceId` in the `Response`
> 
>     6. Use that `deviceId` in your code when calling `Client.get_access_token`
> 
>     7. The response will parrot the `device-id` and give a new `access_token`

@rbrisita Are you using the Venmo client to make any payments? I agree that it works when you follow the steps to get your device ID but the making a payment is the portion. Personally I have trouble with as I think the eligibility token needs to be implemented yet.

**Comment by rbrisita** [comment]
@raycekar I have successfully made charge requests through the API.

**Comment by raycekar** [comment]
@rbrisita could you share me your script (username/password/device_id removed of course).  With my script, I can alter it to include device id and get all the way to the requesting the payment and getting the following error:

`venmo_api.models.exception.HttpCodeError: HTTP Status code is invalid. Could not make the request because -> 403 Forbidden.
Error: {'error': {'message': 'There was an issue with your payment. Try again later.', 'code': 1384, 'links': [], 'title': 'Payment Declined'}}
`

FYI, trying to automate paying with a debit card and it works when manually sending the payment through the app (and I need it to the be the debit card)

**Comment by rbrisita** [comment]
@raycekar From your error, it looks like it's the payment and not the access token. An access token error states:

```json
{
  "error": {
    "message": "You did not pass a valid OAuth access token.",
    "code": 261,
    "links": null,
    "title": "Error"
  }
}
```
I would look into supplying the funding source id if you haven't already for the `send_money` call.
Nothing special about the following code:
```python
from venmo_api import Client
from dotenv import dotenv_values

config = dotenv_values(".env")

user_id = config.get("TARGET_USER_ID", getenv("TARGET_USER_ID"))

username = config.get("VENMO_USER", getenv("VENMO_USER"))
password = config.get("VENMO_PASS", getenv("VENMO_PASS"))
device_id = config.get("VENMO_DEVICE_ID", getenv("VENMO_DEVICE_ID")) # device id taken from a desktop browser

access_token = Client.get_access_token(
  username=username,
  password=password,
  device_id=device_id
)

client = Client(access_token=access_token)

try:
  response = client.payment.request_money(0.01, "Some Message", user_id)
except Exception as e:
  print("Error:", e)

print("Response", response)
```

**Comment by raycekar** [comment]
Thank @rbrisita a ton!
I must had something messed up with my convoluted script but messing around with it from your sample script got me going again

**Comment by rbrisita** [comment]
Awesome! Glad to be of help.

**Comment by jackguo709** [comment]
Question. Is the device ID linked to a particular physical device? If I want to run the python script in a cloud service, then I can't get the device ID ahead of the time

**Comment by afriedsam** [comment]
@jackguo709 I believe you could use any device to generate and save your access token. Then, you could skip the earlier steps and just use `client = Client(access_token=access_token)` in your cloud service.

**Comment by jackguo709** [comment]
But I find that if I try to send multiple payments in a short amount of time, or Venmo someone two or three times, the payments will get denied.

**Comment by rbrisita** [comment]
The `device id` is cross device. Venmo has restrictions on requesting and paying. IIRC, only 30 payment requests per 24 hour period. One can pay as low as 1¢.
